### ISSUE TITLE
Regress in inference of parameters of generic type after 1.1.173

### ISSUE THREAD

**Issue by ethframe** [microsoft/pyright#2445]
**Describe the bug**
After 1.1.173 pyright fails to infer parameters of generic type in some cases.

**To Reproduce**

```
$ npm install -g pyright@1.1.172
...
$ pyright bug.py
...
...\bug.py
  ...\bug.py:21:21 - info: Type of "FnParser(fmap)" is "FnParser[T@Parser, U@fmap]"
0 errors, 0 warnings, 1 info
...
$ npm install -g pyright@1.1.173
...
$ pyright bug.py
...
...\bug.py
  ...\bug.py:21:21 - info: Type of "FnParser(fmap)" is "FnParser[Unknown, Unknown]"
  ...\bug.py:22:16 - error: Return type, "FnParser[Unknown, Unknown]", is partially unknown (reportUnknownVariableType)
1 error, 0 warnings, 1 info
```

**Expected behavior**
No errors.

**Screenshots or Code**

```python
from typing import Callable, Generic, Sequence, TypeVar

T = TypeVar("T")
V = TypeVar("V", bound=object)
V_co = TypeVar("V_co", covariant=True)
U = TypeVar("U", bound=object)


class Result(Generic[V]):
    pass


ParseFn = Callable[[Sequence[T], int, int], Result[V]]


class Parser(Generic[T, V_co]):
    def fmap(self, fn: Callable[[V_co], U]) -> "Parser[T, U]":
        def fmap(stream: Sequence[T], pos: int, bt: int) -> Result[U]:
            raise NotImplementedError()

        return FnParser(fmap)


class FnParser(Parser[T, V_co]):
    def __init__(self, fn: ParseFn[T, V_co]):
        self._fn = fn

    def __call__(self, stream: Sequence[T], pos: int, bt: int) -> Result[V_co]:
        return self._fn(stream, pos, bt)
```

**VS Code extension or command-line**
pyright 1.1.173+
